src/Raven.Client/Http/RequestExecutor.cs

```diff
@@ -1342,6 +1447,83 @@ public static async Task<Stream> ReadAsStreamUncompressedAsync(HttpResponseMessa
             return true;
         }
 
+        private bool ShouldBroadcast<TResult>(RavenCommand<TResult> command)
+        {
+            if (command is IBroadcast == false)
+                return false;
+
+            if (TopologyNodes == null || 
+                TopologyNodes.Count < 2)
+                return false;
+
+            return true;
+        }
+
+        private async Task<TResult> Broadcast<TResult>(RavenCommand<TResult> command, SessionInfo sessionInfo, CancellationToken token)
+        {
+            var broadcastCommand = command as IBroadcast;
+            if (broadcastCommand == null)
+                throw new InvalidOperationException("You can broadcast only commands that implement 'IBroadcast'.");
+
+            command.FailedNodes = new Dictionary<ServerNode, Exception>(); // clear the current failures
+
+            using (var broadcastCts = CancellationTokenSource.CreateLinkedTokenSource(token))
+            {
+                var tasks = SendToAllNodes<TResult>(sessionInfo, broadcastCommand, broadcastCts.Token);
+
+                return await WaitForResult(command, tasks, broadcastCts).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<TResult> WaitForResult<TResult>(RavenCommand<TResult> command, Dictionary<Task, (RavenCommand<TResult> Command, int Index)> tasks, CancellationTokenSource broadcastCts)
+        {
+            while (tasks.Count > 0)
+            {
+                var completed = await Task.WhenAny(tasks.Keys).ConfigureAwait(false);
+                if (completed.IsCanceled || completed.IsFaulted)
+                {
+                    var failed = tasks[completed];
+                    var node = _nodeSelector.Topology.Nodes[failed.Index];
+                    command.FailedNodes[node] = completed.Exception.ExtractSingleInnerException();
+                    _nodeSelector.OnFailedRequest(failed.Index);
+
+                    tasks.Remove(completed);
+                    continue;
+                }
+
+                broadcastCts.Cancel(throwOnFirstException: false);
+
+                _nodeSelector.RestoreNodeIndex(tasks[completed].Index);
+                return tasks[completed].Command.Result;
+            }
+
+            var ae = new AggregateException(command.FailedNodes.Select(x => new UnsuccessfulRequestException(x.Key.Url, x.Value)));
+            throw new AllTopologyNodesDownException($"Broadcasting {command.GetType()} failed.", ae);
+        }
+
+        private Dictionary<Task, (RavenCommand<TResult> Command, int Index)> SendToAllNodes<TResult>(SessionInfo sessionInfo, IBroadcast command, CancellationToken token)
+        {
```

Comment:
if this method will throw we will leak the context, maybe we should have a try/catch clause here and in case of an exception we should cancel everything and return all contexts?

Reply:
done